spec/lib/bolt_spec/plans/mock_executor.rb

```diff
@@ -0,0 +1,196 @@
+          val = @data[target.name] || @data[:default]
+          Bolt::Result.new(target, value: val)
+        end)
+      end
+
+      def assert_called(taskname)
+        satisfied = if @expect
+                      (@expected_calls.nil? && @calls > 0) || @calls == @expected_calls
+                    else
+                      @expected_calls.nil? || @calls <= @expected_calls
+                    end
+        unless satisfied
+          unless (times = @expected_calls)
+            times = @expect ? "at least one" : "any number of"
+          end
+          message = "Expected #{taskname} to be called #{times} times"
+          message += " with targets #{@invocation[:targets]}" if @invocation[:targets]
+          message += " with parameters #{@invocations[:parameters]}" if @invocation[:parameters]
+          raise message
+        end
+      end
+
+      # This changes the stub from an allow to an expect which will validate
+      # that it has been called.
+      def expect_call
+        @expect = true
+        self
+      end
+
+      # Below here are the intended 'public' functions
+
+      # Restricts the stub to only match invocations with
+      # the correct targets
+      def with_targets(targets)
+        targets = [targets] unless targets.is_a? Array
+        @invocation[:targets] = targets.map do |target|
+          if target.is_a? String
+            target
+          else
+            target.name
+          end
+        end
+        self
+      end
+
+      # Restricts the stub to only match invocations with certain parameters
+      # All parameters must match exactly and since arguments and options are
+      # treated differently at the executor this won't work with some '_*' options
+      # TODO: Fix handling of '_*' options probably by breaking them into other helpers
+      def with_params(params)
+        @invocation[:parameters] = params
+        @invocation[:arguments] = params.reject { |k, _v| k.start_with?('_') }
+        @invocation[:options] = params.select { |k, _v| k.start_with?('_') }
+        self
+      end
+
+      # limit the maximum number of times an allow stub may be called or
+      # specify how many times an expect stub must be called.
+      def be_called_times(times)
+        @expected_calls = times
+        self
+      end
+
+      # error if the stub is called at all.
+      def not_be_called
+        @expected_calls = 0
+        self
+      end
+
+      # Set different result values for each target
+      # TODO: how should this interact with multiple calls and always_return
+      def return_for_targets(data)
+        data.each do |target, result|
+          raise "Mocked results must be hashes: #{target}: #{result}" unless result.is_a? Hash
+        end
+        @data = @data.merge(data)
+        self
+      end
+
+      # Set a default return value for all targets, specific targets may be overrided with return_for_targets
+      def always_return(default_data)
```

Comment:
As suggested above, the return value should not be tied to the task this tightly, the task should be just another constraint on the stub just like any of the other parameters. I.e. the return value should only be tied to a particular stub, whether it is used or not should only depend on whether the constraints on the stub match the passed parameters (including the target(s)) or not.

Reply:
I'm not sure what the use case is for stubbing some but not all tasks at the same time. I think at that point you should probably just write your own mock executor.

matching by task first makes it less likely to have to worry about conflicts. I could see making the "default" double match on tasks and perhaps get called as a fall back if there are no task double matches